### PR TITLE
Adding $nor operator support 

### PR DESCRIPTION
**Pull request by jeremypiednoel**
@dplewis tell me if you want me to investigate on postgresql !

### CODE REVIEW

**Comment by dplewis** [comment]
@jeremypiednoel Go for it! Let me know if you need anything.

**Comment by jeremypiednoel** [comment]
@dplewis I'm very noob in postgresql. So let's see if i can figure out

**Comment by flovilmart** [comment]
Very nice! We'll need to add some docs as the logical explanation is quite complex :)

**Comment by jeremypiednoel** [comment]
Btw am i the only one to have pb with `.github/ISSUE_TEMPLATE/Bug_report.md` maybe it's my IDE but it always update this file and ask me to commit it

**Comment by flovilmart** [comment]
@jeremypiednoel did you pull the latest master& this should be good now.

**Comment by dplewis** [comment]
@jeremypiednoel That issue was fixed minutes ago https://github.com/parse-community/parse-server/pull/4767

**Comment by jeremypiednoel** [comment]
@dplewis very true i had to commit the file first to pull the upstream then.
nvm it's fixed 👍

**Comment by jeremypiednoel** [comment]
@dplewis how can i specify i want the test to run on postgresql ?

**Comment by dplewis** [comment]
Check `CONTRIBUTING.md`

https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md

**Comment by jeremypiednoel** [comment]
Thanks ! i'll try to make it work

**Comment by dplewis** [comment]
@jeremypiednoel If you want to cheat you can find a solution on my branch for PG. 

https://github.com/dplewis/parse-server/tree/nor-op

You will still need test for empty $nor array (I know that mongo and pg throws an error for nonempty $or, $and, $nor) and a test for invalid nor (not array)

**Comment by jeremypiednoel** [comment]
thnaks @dplewis i'm gonna check that after lunch !

**Comment by jeremypiednoel** [comment]
@dplewis can you take a look ?

File: spec/ParseQuery.spec.js [hunk 1]
```diff
@@ -2567,6 +2567,48 @@ describe('Parse.Query testing', () => {
     });
   });
 
+  it('$select inside $nor', (done) => {
```
Comment (dplewis) [inline spec/ParseQuery.spec.js]: Change the name as `$select` isn't being used here

File: spec/ParseQuery.spec.js [hunk 2]
```diff
@@ -2567,6 +2567,48 @@ describe('Parse.Query testing', () => {
...
+    const options = Object.assign({}, masterKeyOptions, {
+      body: {
+        where: { $nor: [] },
+      }
+    });
+    const obj = new TestObject();
+    obj.save().then(() => {
+      return rp.get(Parse.serverURL + "/classes/TestObject", options);
+    }).then(done.fail).catch((error) => {
+      equal(error.error.code, Parse.Error.INVALID_JSON);
```
Comment (dplewis) [inline spec/ParseQuery.spec.js]: Should be `INVALID_QUERY`.

File: src/Controllers/DatabaseController.js [hunk 3]
```diff
@@ -62,7 +62,7 @@ const validateQuery = (query: any): void => {
   }
 
   if (query.$or) {
-    if (query.$or instanceof Array) {
```
Comment (dplewis) [inline src/Controllers/DatabaseController.js]: Revert these changes to `$or` and `$and` as are they are causing tests to fail.

They also have nothing to do with the PR.

File: spec/ParseQuery.spec.js [hunk 4]
```diff
@@ -2567,6 +2567,48 @@ describe('Parse.Query testing', () => {
...
+    Parse.Object.saveAll(objects).then(() => {
+      return rp.get(Parse.serverURL + "/classes/TestObject", options);
+    }).then((results) => {
+      expect(results.results.length).toBe(highValue - lowValue);
+      expect(results.results.every(res => res.rating > lowValue && res.rating <= highValue)).toBe(true);
+      done();
+    });
+  });
+
+  it('$nor invalid query', (done) => {
```
Comment (dplewis) [inline spec/ParseQuery.spec.js]: Maybe another test when `$nor: 1234` (non array)

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 5]
```diff
@@ -306,12 +306,15 @@ const buildWhereClause = ({ schema, query, index }): WhereClause => {
...
-    } else if (fieldName === '$or' || fieldName === '$and') {
+    } else if (['$or', '$nor', '$and'].includes(fieldName)) {
       const clauses = [];
       const clauseValues = [];
       fieldValue.forEach((subQuery) =>  {
         const clause = buildWhereClause({ schema, query: subQuery, index });
         if (clause.pattern.length > 0) {
+          if (fieldName === '$nor') {
+            clause.pattern = `(NOT ${clause.pattern})`;
+          }
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: you're doing `(NOT query1) AND (NOT query2)` instead of `NOT (query1 OR query2)` whhich tecnically work but can be a bit misleading for maintenance. Cna you document it?

Comment (dplewis) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: True I was pretty lazy 😅

Comment (flovilmart) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Yeah I figured it worked but letMs write it cleanly please

Comment (jeremypiednoel) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: I'll fix that

**Comment by jeremypiednoel** [comment]
@flovilmart @dplewis i reverted the logic.
Thanks for the feedbacks

**Comment by flovilmart** [comment]
@jeremypiednoel thanks for the changes, looking good, waiting for the CI.

**Comment by jeremypiednoel** [comment]
@flovilmart do you think it makes sense to add the array.length test on $end and $or conditions ?
